pkg/finalize_bundle.go

```diff
@@ -0,0 +1,165 @@
+package cnab_riff
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strings"
+
+	"github.com/deislabs/cnab-go/bundle"
+	"github.com/deislabs/duffle/pkg/duffle/manifest"
+	"github.com/ghodss/yaml"
+	"github.com/pivotal/image-relocation/pkg/image"
+	"github.com/pivotal/image-relocation/pkg/registry"
+	"github.com/projectriff/cnab-k8s-installer-base/pkg/apis/kab/v1alpha1"
+	"github.com/projectriff/k8s-manifest-scanner/pkg/scan"
+)
+
+// this performs following tasks:
+// 1. inlines the content of the resource url into the bundle
+// 2. adds images to duffle.json by scanning the resource content
+// 3. computes digests for images
+// 4. replaces image references in kab manifest with digested references
+func FinalizeBundle(bundlePath, kabManifestPath, kabManifestDestinationPath string) error {
+	mfst := &manifest.Manifest{}
+	err := unmarshallFile(bundlePath, mfst)
+	if err != nil {
+		return err
+	}
+
+	kabMfst := &v1alpha1.Manifest{}
+	err = unmarshallFile(kabManifestPath, kabMfst)
+	if err != nil {
+		return err
+	}
+
+	err = kabMfst.InlineContent()
+	if err != nil {
+		return err
+	}
+
+	images, err := GetImagesFromKabManifest(kabMfst)
+	if err != nil {
+		return err
+	}
+
+	mfst.Images = map[string]bundle.Image{}
+	registryClient := registry.NewRegistryClient()
+	replacements := []string{}
+
+	for _, img := range images {
+		name, err := image.NewName(img)
+		if err != nil {
+			fmt.Printf("err %v\n", err)
+		}
+		bundleImageKey := strings.ReplaceAll(name.String(), "/", "_")
+		bundleImage := bundle.Image{}
+
+		digest, err := registryClient.Digest(name)
+		if err != nil {
+			return err
+		}
+		bundleImage.Digest = digest.String()
+		nameWithDigest, err := name.WithDigest(digest)
+		if err != nil {
+			return err
+		}
+		bundleImage.Image = nameWithDigest.String()
+
+		mfst.Images[bundleImageKey] = bundleImage
+
+		replacements = append(replacements, img, nameWithDigest.String())
+	}
+
+	err = marshallJsonFile(bundlePath, mfst)
+	if err != nil {
+		return err
+	}
+
+	err = ReplaceInKabManifest(kabMfst, *strings.NewReplacer(replacements...))
+	if err != nil {
+		return err
+	}
+
+	err = marshallYamlFile(kabManifestDestinationPath, kabMfst)
+	return err
+
+}
+
+func unmarshallFile(path string, str interface{}) error {
```

Comment:
Best to adopt US spelling for consistency with the Go standard library.
```suggestion
func unmarshalFile(path string, str interface{}) error {
```